### PR TITLE
search: enrich lucky search suggestions

### PR DESCRIPTION
**Pull request by rvantonder**
lucky search currently notifies the user of alternative queries that ran, even when those queries return 0 results. This PR collects the count of results coming from each generated query and only sends the notification if we found results.

Screen Shot 2022-06-23 at 12 39 32 PM

Pretty cool! Yes, this needs more bookkeeping overhead but it's so valuable I think it's worth it. The previous state is OK for a small number of suggestions, but as the query generation grows, this is going to get out of hand.

Aside: generated query job cancellation appears to not work, it's supposed to also not notify if we've already sent 500+ results total. I've never seen the context cancelled abort lucky search [in this code block locally](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/internal/search/job/jobutil/feeling_lucky_search_job.go?L87-92), and I've tried to force it to. This is a separate issue to investigate and shouldn't block this PR though.

## Test plan
Experimental and will add a test, putting up for early review/feedback.

### CODE REVIEW

File: internal/search/job/jobutil/feeling_lucky_search_job.go [hunk 1]
```diff
@@ -462,8 +465,31 @@ type generatedSearchJob struct {
...
-func (g *generatedSearchJob) Run(ctx context.Context, clients job.RuntimeClients, stream streaming.Sender) (*search.Alert, error) {
+func (g *generatedSearchJob) Run(ctx context.Context, clients job.RuntimeClients, parentStream streaming.Sender) (*search.Alert, error) {
+	var resultCount int
+	var mux sync.Mutex
+	stream := streaming.StreamFunc(func(event streaming.SearchEvent) {
+		mux.Lock()
+		resultCount += event.Results.ResultCount()
+		mux.Unlock()
+		parentStream.Send(event)
+	})
```
Comment (camdencheek) [inline internal/search/job/jobutil/feeling_lucky_search_job.go:476]: check out [NewResultCountingStream](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@d8eb966482b5d0db2a0c6d13b14eeb3acc749cbb/-/blob/internal/search/streaming/stream.go?L170-174)

**Review by camdencheek — APPROVED** [review]

**Comment by camdencheek** [comment]
> generated query job cancellation appears to not work

That's not great. I'll take a look at that since we depend on that in a lot of places

File: internal/search/job/jobutil/feeling_lucky_search_job.go [hunk 2]
```diff
@@ -462,8 +465,31 @@ type generatedSearchJob struct {
...
+	resultCountString := strconv.Itoa(resultCount)
+	if resultCount == limits.DefaultMaxSearchResultsStreaming {
+		resultCountString = fmt.Sprintf("%d+ results", resultCount)
+	} else if resultCount == 1 {
+		resultCountString = fmt.Sprintf("1 result")
+	} else {
+		resultCountString = fmt.Sprintf("%d results", resultCount)
+	}
+
+	g.ProposedQuery.Description = fmt.Sprintf("%s (%s)", g.ProposedQuery.Description, resultCountString)
```
[the remark this answers is not in the document]
Comment (rvantonder) [reply internal/search/job/jobutil/feeling_lucky_search_job.go:492]: Yeah, I remember :P I'll try, it's tricky, gotta detangle query generation and description. It's that whole "difficult to cleanly separate information across entire end-to-end behavior"

File: internal/search/job/jobutil/feeling_lucky_search_job.go (hunk 1 above)
Comment (rvantonder) [reply internal/search/job/jobutil/feeling_lucky_search_job.go:476]: this is actually fricken amazing. finally having the right things to reuse

Comment (camdencheek) [reply internal/search/job/jobutil/feeling_lucky_search_job.go:476]: composable streams + composable jobs = 💯

**Comment by rvantonder** [comment]
> That's not great. I'll take a look at that since we depend on that in a lot of places

Yeah it's weird because I tested this behavior in SequentialJob:
https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/internal/search/job/jobutil/combinators_test.go?L205. So I'm guessing it's not the logic that's actually broken, just that we don't see something propagated? Dunno

**Comment by rvantonder** [comment]
@camdencheek actually I think I see the problem with the cancellation now that I linked that test. `LuckySearchJob` isn't wrapped in any limit: https://sourcegraph.com/github.com/sourcegraph/sourcegraph@6d0ac006169ffde338a85b6a1421dc08f2a23779/-/blob/internal/search/job/jobutil/job.go?L46. Looks like putting a `NewLimitJob()` around it fixes it! Will confirm a bit later.

**Comment by camdencheek** [comment]
> Looks like putting a NewLimitJob() around it fixes it! Will confirm a bit later.

ope, that'll do it! Good catch 👍

**Pull request merged**